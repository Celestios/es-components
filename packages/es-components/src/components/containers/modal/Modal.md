Use the `closeButton` attribute on the Modal.Header to show or hide the "X" button.

For accessbility, make sure to include the attribute `aria-haspopup='dialog'` on whatever link you create to activate the modal. If you need
an informational dialog that contains no focusable elements, use a popover or something else more appropriate. Additionally, you can supply
a level prop to the `Modal.Header` to prevent skipping heading levels. The style of the `Modal.Header` will remain the same regardless of
the level passed in.

```
import Button from '../../controls/buttons/Button';

const [state, setState] = React.useState({show: false});

<div>
  <Button aria-haspopup='dialog' onClick={() => setState({show: true, size: 'small', hideCloseButton: true})} style={{marginRight:'15px'}}>Open Small Modal</Button>

  <Button aria-haspopup='dialog' onClick={() => setState({show: true, size: 'medium', hideCloseButton: false})} style={{marginRight:'15px'}}>Open Medium Modal</Button>

  <Button aria-haspopup='dialog' onClick={() => setState({show: true, size: 'large', hideCloseButton: false})}>Open Large Modal</Button>

  <Modal
    size={state.size}
    show={state.show}
    onHide={() => setState({show: false})}
  >
    <Modal.Header level={4} hideCloseButton={state.hideCloseButton}>This is the header.</Modal.Header>
    <Modal.Body>Body Content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch.This is the popover's content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid.</Modal.Body>
    <Modal.Footer>
      <span style={{ flexGrow: 1, marginRight: '10px' }}>
        This is the footer.
      </span>
      <Button onClick={()=> setState({show: false})} style={{margin:0}}>Ok</Button>
    </Modal.Footer>
  </Modal>
</div>
```

`ModalButtonContainer` is a provided wrapper to give consistent styling to buttons and their layout inside of `Modal`s.

```
import ModalButtonContainer from './ModalButtonContainer';
import Button from '../../controls/buttons/Button';

const [state, setState] = React.useState({show: false});

<div>
  <Button aria-haspopup='dialog' onClick={() => setState({show: true, size: 'medium', hideCloseButton: false})} style={{marginRight:'15px'}}>Open Modal</Button>

  <Modal
    size="medium"
    show={state.show}
    onHide={() => setState({show: false})}
  >
    <Modal.Header level={4} hideCloseButton={state.hideCloseButton}>This is the header.</Modal.Header>
    <Modal.Body>Body Content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch.This is the popover's content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid.</Modal.Body>
    <Modal.Footer>
      <ModalButtonContainer>
        <Button onClick={()=> setState({show: false})} style={{margin: '15px 0 0 15px'}}>Cancel</Button>
        <Button onClick={()=> setState({show: false})} styleType="primary" style={{margin: '15px 0 0 15px'}}>Ok</Button>
      </ModalButtonContainer>
    </Modal.Footer>
  </Modal>
</div>
```
