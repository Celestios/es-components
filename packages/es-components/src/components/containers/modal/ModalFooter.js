import styled from 'styled-components';

// Note: ModalFooter relies on a parent (Modal) with ThemeProvider wrapping it
const ModalFooter = styled.div`
  display: flex;
  padding: 15px;
  text-align: right;

  button {
    width: auto;
  }
`;

export default ModalFooter;
