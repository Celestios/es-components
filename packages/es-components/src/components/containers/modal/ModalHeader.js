import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import DismissButton from '../../controls/DismissButton';
import { ModalContext } from './ModalContext';
import Heading from '../heading/Heading';

// Note: ModalHeader relies on a parent (Modal) with ThemeProvider wrapping it
const Header = styled.div`
  border-bottom: 2px solid ${props => props.theme.brandColors.vbMagenta};
  color: ${props => props.theme.colors.gray9};
  display: flex;
  font-size: 26px;
  font-weight: 500;
  justify-content: space-between;
  padding: 15px;
`;

const Title = styled(Heading)`
  font-size: 24px;
  font-weight: 500;
  margin: 0;
`;

const DismissModal = styled(DismissButton)`
  color: ${props => props.theme.colors.black};
  line-height: 1;
  opacity: 0.2;
  text-shadow: 0 1px 0 ${props => props.theme.colors.white};

  &:hover {
    opacity: 0.5;
  }
`;

function ModalHeader(props) {
  const { hideCloseButton, children, level, ...otherProps } = props;
  const { onHide, ariaId } = useContext(ModalContext);

  return (
    <Header {...otherProps}>
      <Title id={ariaId} level={level}>
        {children}
      </Title>

      {!hideCloseButton && <DismissModal onClick={onHide} />}
    </Header>
  );
}

ModalHeader.propTypes = {
  /** Specify whether the modal header should contain a close button */
  hideCloseButton: PropTypes.bool,
  children: PropTypes.any,
  level: PropTypes.oneOf([1, 2, 3, 4, 5, 6])
};

ModalHeader.defaultProps = {
  hideCloseButton: false,
  children: undefined,
  level: 4
};

ModalHeader.contextTypes = {
  modal: PropTypes.shape({
    onHide: PropTypes.func,
    ariaId: PropTypes.string
  })
};

export default ModalHeader;
