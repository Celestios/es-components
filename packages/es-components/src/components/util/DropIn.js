import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const DropIn = ({ children, duration, ...otherProps }) => {
  const transitionStyles = {
    opacity: 0,
    transform: `translate(0, -25%)`,
    transition: `transform ${duration}ms ease-out, opacity ${duration}ms linear`
  };

  const stateStyles = {
    entering: { opacity: 1, transform: `translate(0, 0)` },
    entered: { opacity: 1, transform: `translate(0, 0)` }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

DropIn.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number
};

DropIn.defaultProps = {
  duration: 300
};

export default DropIn;
