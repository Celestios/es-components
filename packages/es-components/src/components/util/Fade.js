import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const Fade = ({ children, duration, opacity, ...otherProps }) => {
  const transitionStyles = {
    transition: `opacity ${duration}ms linear`,
    opacity: 0
  };

  const stateStyles = {
    entering: { opacity },
    entered: { opacity }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
