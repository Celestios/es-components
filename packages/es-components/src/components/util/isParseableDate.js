import parse from 'date-fns/parse';
import isValid from 'date-fns/isValid';

export default function isParseableDate(date) {
  const parsedDate = parse(date);
  return isValid(parsedDate);
}
