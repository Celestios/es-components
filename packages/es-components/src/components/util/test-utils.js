// eslint-env jest
import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function renderWithTheme(component) {
  return render(
    <ThemeProvider theme={viaTheme}>
      {React.cloneElement(component)}
    </ThemeProvider>
  );
}
