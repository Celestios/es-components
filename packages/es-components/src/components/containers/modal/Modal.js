import React, { useCallback, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { createGlobalStyle, ThemeProvider } from 'styled-components';
import { noop, debounce } from 'lodash';
import ReactModal from 'react-modal';
import tinycolor from 'tinycolor2';

import useUniqueId from '../../util/useUniqueId';
import { useRootNodeLocator } from '../../util/useRootNode';
import { ModalContext } from './ModalContext';
import Header from './ModalHeader';
import Body from './ModalBody';
import Footer from './ModalFooter';

const modalSize = {
  small: '300px',
  medium: '600px',
  large: '900px'
};

const animationTimeMs = 300;

const getModalMargin = (windowHeight, modalHeight) => {
  const fullMargin = windowHeight - modalHeight;
  const thirdTopMargin = fullMargin / 3;
  return Math.max(thirdTopMargin, 0);
};

const ModalStyles = createGlobalStyle`
  body.modal-open {
    overflow: hidden;

    .background-overlay {
      bottom: 0;
      left: 0;
      position: fixed;
      right: 0;
      top: 0;
      background-color: ${props => {
        const color = tinycolor(props.theme.colors.black);
        color.setAlpha(0);
        return color.toRgbString();
      }};
      transition: background-color ${() => animationTimeMs}ms linear;
      z-index: 1030;

      &.ReactModal__Overlay--after-open {
        ${props => {
          if (!props.showBackdrop) return '';

          const color = tinycolor(props.theme.colors.black);
          color.setAlpha(0.5);
          return `background-color: ${color.toRgbString()};`;
        }}

        &.ReactModal__Overlay--before-close {
          background-color: ${props => {
            const color = tinycolor(props.theme.colors.black);
            color.setAlpha(0);
            return color.toRgbString();
          }};
        }
      }
    }

    .modal-content {
      background-clip: padding-box;
      background-color: ${props => props.theme.colors.white};
      border-radius: 3px;
      box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
      font-family: 'Source Sans Pro', 'Segoe UI', Segoe, Calibri, Tahoma, sans-serif;
      font-size: ${props => props.theme.font.baseFontSize};
      outline: 0;
      opacity: 0;
      overflow: scroll;
      position: relative;
      transition:
        top ${() => animationTimeMs}ms ease-out,
        opacity ${() => animationTimeMs}ms linear;
      -webkit-overflow-scrolling: touch;
      width: 100%;
      max-height: 100%;
      ${props =>
        props.showAnimation
          ? `
            top: -100%;
          `
          : ''};

      &.ReactModal__Content--after-open {
        ${props =>
          props.showAnimation
            ? `
              opacity: 1;
              top: 0;
            `
            : ''};

        &.ReactModal__Content--before-close {
          ${props =>
            props.showAnimation
              ? `
                opacity: 0;
                top: -100%
              `
              : ''};
        }
      }

      &.small {
        @media (min-width: ${props => props.theme.screenSize.phone}) {
          margin: ${({ theme }) =>
            getModalMargin(theme.windowHeight, theme.modalHeight)}px auto;
          width: ${modalSize.small};
        }
      }

      &.medium {
        @media (min-width: ${props => props.theme.screenSize.tablet}) {
          margin: ${({ theme }) =>
            getModalMargin(theme.windowHeight, theme.modalHeight)}px auto;
          width: ${modalSize.medium};
        }
      }

      &.large {
        @media (min-width: ${props => props.theme.screenSize.desktop}) {
          margin: ${({ theme }) =>
            getModalMargin(theme.windowHeight, theme.modalHeight)}px auto;
          width: ${modalSize.large};
        }
      }
    }
  }
`;

function Modal({
  animation,
  backdrop,
  children,
  escapeExits,
  onEnter,
  onExit,
  onHide,
  show,
  size,
  parentSelector,
  className,
  contentRef,
  ...other
}) {
  const ariaId = useUniqueId(other.id);
  const [rootNode, RootNodeLocator] = useRootNodeLocator(document.body);
  const [headerHeight, setHeaderHeight] = useState(60);
  const [modalHeight, setModalHeight] = useState(300);
  const [shouldShow, setShouldShow] = useState(show);
  const [windowHeight, setWindowHeight] = useState(window.innerHeight);

  const modalHeightRef = useCallback(modalElement => {
    if (contentRef) {
      if (typeof contentRef === 'function') {
        contentRef(modalElement);
      } else {
        // eslint-disable-next-line no-param-reassign
        contentRef.current = modalElement;
      }
    }

    setModalHeight(modalElement?.offsetHeight || 0);
  });
  const modalParentSelector =
    parentSelector || useCallback(() => rootNode, [rootNode]);

  useEffect(() => {
    if (!animation || show) {
      setShouldShow(show);
      return noop;
    }

    let mounted = true;
    setTimeout(() => {
      if (!mounted) return;

      setShouldShow(false);
    }, animationTimeMs);

    return () => {
      mounted = false;
    };
  }, [show, animation]);

  useEffect(() => {
    const setHeight = debounce(() => setWindowHeight(window.innerHeight), 100);
    window.addEventListener('resize', setHeight);
    return () => window.removeEventListener('resize', setHeight);
  }, []);

  const shouldCloseOnOverlayClick = backdrop !== 'static' && backdrop;

  return (
    <ThemeProvider
      theme={{ headerHeight, setHeaderHeight, modalHeight, windowHeight }}
    >
      <RootNodeLocator />
      {shouldShow ? (
        <ModalStyles showAnimation={animation} showBackdrop={backdrop} />
      ) : null}
      <ModalContext.Provider value={{ onHide, ariaId }}>
        <ReactModal
          bodyOpenClassName="modal-open"
          className={`${className} ${size} modal-content`}
          overlayClassName="background-overlay"
          closeTimeoutMS={animation ? animationTimeMs : null}
          isOpen={show}
          aria={{
            labelledby: ariaId
          }}
          shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
          onRequestClose={onHide}
          onAfterOpen={onEnter}
          onAfterClose={onExit}
          shouldCloseOnEsc={escapeExits}
          parentSelector={modalParentSelector}
          contentRef={modalHeightRef}
          {...other}
        >
          {children}
        </ReactModal>
      </ModalContext.Provider>
    </ThemeProvider>
  );
}

Modal.propTypes = {
  /** Open and close the Modal with transitions. */
  animation: PropTypes.bool,
  /**
   * Include a backdrop component. Specify 'static' for a backdrop that doesn't
   * trigger an "onHide" when clicked.
   */
  backdrop: PropTypes.oneOf(['static', true, false]),
  /**
   * Usually contains a Modal.Header, Modal.Body, and Modal.Footer
   * but can contain any content
   */
  children: PropTypes.any,
  /** Set whether the Escape key exits the modal */
  escapeExits: PropTypes.bool,
  /** Callback fired when the Modal transitions in */
  onEnter: PropTypes.func,
  /** Callback fired when the modal transitions out */
  onExit: PropTypes.func,
  /**
   * A callback fired when the header closeButton or non-static backdrop is
   * clicked. Required if either are specified.
   */
  onHide: PropTypes.func,
  /** When `true` The modal will show itself. */
  show: PropTypes.bool,
  /** Sets the size of the modal */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /** Selects the parent */
  parentSelector: PropTypes.func,
  className: PropTypes.string,
  contentRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any })
  ])
};

Modal.defaultProps = {
  animation: true,
  backdrop: true,
  escapeExits: true,
  onEnter: noop,
  onExit: noop,
  onHide: noop,
  show: false,
  size: 'medium',
  children: undefined,
  parentSelector: null,
  className: null,
  contentRef: null
};

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

export default Modal;
