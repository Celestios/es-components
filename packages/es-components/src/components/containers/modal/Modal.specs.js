/* eslint-env jest */

import React from 'react';
import { fireEvent, cleanup, wait } from 'react-testing-library';

import Modal from './Modal';
import ModalHeader from './ModalHeader';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';
import { renderWithTheme } from '../../util/test-utils';

jest.mock('../../util/generateAlphaName', () => () => 'abcdef');

beforeEach(cleanup);

function renderModal(modalProps) {
  return renderWithTheme(
    <Modal show {...modalProps}>
      <ModalHeader>Header</ModalHeader>
      <ModalBody>Body</ModalBody>
      <ModalFooter>Footer</ModalFooter>
    </Modal>
  );
}

it('invokes functions when triggered', () => {
  const onEnter = jest.fn();
  const onHide = jest.fn();

  const { getByText } = renderModal({ onEnter, onHide });

  expect(onEnter).toHaveBeenCalled();

  getByText('Header')
    .parentElement.querySelector('button')
    .click();
  expect(onHide).toHaveBeenCalled();
});

it('renders different modal sections', () => {
  const onEnter = jest.fn();
  const onHide = jest.fn();

  const { getByText } = renderModal({ onEnter, onHide });

  expect(onEnter).toHaveBeenCalled();

  expect(getByText('Header')).toMatchSnapshot();
  expect(getByText('Body')).toMatchSnapshot();
  expect(getByText('Footer')).toMatchSnapshot();
});

it('hides dismiss button when hideCloseButton is true', () => {
  const { getByText } = renderWithTheme(
    <Modal show>
      <ModalHeader hideCloseButton>Header</ModalHeader>
      <ModalBody>Body</ModalBody>
      <ModalFooter>Footer</ModalFooter>
    </Modal>
  );
  expect(getByText('Header').parentElement.querySelector('button')).toBeNull();
});

it('passes through class names', () => {
  const { container } = renderModal({ className: 'myclass' });
  const element = container.parentElement.querySelector('.myclass');
  expect(element).not.toBeNull();
});

describe('when ESC is pressed', () => {
  it('invokes onHide by default', () => {
    const onHide = jest.fn();
    const { container } = renderModal({ onEnter: jest.fn(), onHide });

    fireEvent.keyDown(container, { keyCode: 27 });
    wait(() => {
      expect(onHide).toHaveBeenCalled();
    });
  });

  it('does not invoke onHide when escapeExits is false', () => {
    const onHide = jest.fn();
    const { container } = renderModal({ onHide, escapeExits: false });

    fireEvent.keyDown(container, { keyCode: 27 });
    wait(() => {
      expect(onHide).not.toHaveBeenCalled();
    });
  });
});
