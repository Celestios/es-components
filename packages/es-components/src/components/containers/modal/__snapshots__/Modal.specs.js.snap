// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders different modal sections 1`] = `
<h4
  class="sc-ifAKCX fBKnDt sc-htpNat kauQMr"
  id="abcdef"
>
  Header
</h4>
`;

exports[`renders different modal sections 2`] = `
<div
  class="sc-bZQynM eVbtnu"
>
  Body
</div>
`;

exports[`renders different modal sections 3`] = `
<div
  class="sc-gzVnrw ilnIUq"
>
  Footer
</div>
`;
